### PR TITLE
Add markov chain based stories to jim

### PR DESCRIPTION
**Pull request by Gagogaa**

### CODE REVIEW

**Comment by markzz** [comment]
Does the book actually need to be part of this repository? I would think it would be better to just have a directory specified and just have Jim look in that directory to see if there's anything there.

File: jim/cmd_funcs.py [hunk 1]
```diff
@@ -10,6 +10,10 @@
 from jim.util import util, DB
 
 
+STORY_TIME_MARKOV = Markov(_stop_condition)
+READ_BOOKS = []
```
Comment (markzz) [inline jim/cmd_funcs.py]: Just a thought, perhaps in this case it doesn't make sense, but for other globals in this code, they are housed in `util.py`. Should these perhaps be housed there as well?

File: jim/cmd_funcs.py [hunk 2]
```diff
@@ -44,6 +48,18 @@
         json.dump(log, f)
 
 
+def _stop_condition(words):
+    # Set a hard limit at 50 so this cannot run forever
+    if len(words) > 50:
+        words[0] = words[0].capitalize()
+        return True
+    # End the chain if its greater than 15 words and ends in '.',  '?' or '!'
+    elif len(words) > 15 and ('.' in words[-1]) or ('?' in words[-1]) or ('!' in words[-1]):
```
Comment (markzz) [inline jim/cmd_funcs.py]: The parenthesis here are redundant and unneeded.

File: jim/cmd_funcs.py [hunk 3]
```diff
@@ -44,6 +48,18 @@
...
+        words[0] = words[0].capitalize()
+        return True
+    return False
```
Comment (markzz) [inline jim/cmd_funcs.py:61]: Style: return should be in its own code group

File: jim/cmd_funcs.py [hunk 4]
```diff
@@ -256,6 +272,39 @@
...
+    books_dir = config.config_get('story', 'books_dir')
+    for bookname in [file for file in os.scandir(books_dir) if file.is_file()]:
+        if bookname not in READ_BOOKS:
+            with open(bookname, 'r') as fin:
+                source = fin.read()
+                words = [word for word in source.translate(translation_table).split(' ')
+                         if word is not ' ' and word is not '']
+                READ_BOOKS.append(bookname)
+                STORY_TIME_MARKOV.learn(words)
+    return ' '.join(next(m.chains()))
```
Comment (markzz) [inline jim/cmd_funcs.py]: Style: return should be in its own code group

File: jim/cmd_funcs.py [hunk 5]
```diff
@@ -44,6 +48,18 @@
         json.dump(log, f)
 
 
+def _stop_condition(words):
+    # Set a hard limit at 50 so this cannot run forever
+    if len(words) > 50:
+        words[0] = words[0].capitalize()
+        return True
+    # End the chain if its greater than 15 words and ends in '.',  '?' or '!'
```
Comment (markzz) [inline jim/cmd_funcs.py]: Move this comment to within the elif block

File: jim/cmd_funcs.py [hunk 6]
```diff
@@ -256,6 +272,39 @@
...
+    books_dir = config.config_get('story', 'books_dir')
+    for bookname in [file for file in os.scandir(books_dir) if file.is_file()]:
+        if bookname not in READ_BOOKS:
+            with open(bookname, 'r') as fin:
+                source = fin.read()
+                words = [word for word in source.translate(translation_table).split(' ')
+                         if word is not ' ' and word is not '']
+                READ_BOOKS.append(bookname)
+                STORY_TIME_MARKOV.learn(words)
+    return ' '.join(next(STORY_TIME_MARKOV.chains()))
```
Comment (markzz) [inline jim/cmd_funcs.py]: Style: return should be in its own code group

File: jim/markov.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+    def __init__(self, stop_condition, starting_symbols=None, memory_bytes=None):
+        if memory_bytes:
+            self._memory = pickle.loads(starting_memory)
+        else:
+            self._memory = defaultdict(list)
+        self._stop_condition = stop_condition
+        self._starting_symbols = starting_symbols
+
+
```
Comment (markzz) [inline jim/markov.py]: PEP-8: Too many blank lines.

File: jim/markov.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+            symbol = self._get_starting_symbol()
+            chain = [symbol]
+            while not self._stop_condition(chain):
+                symbol = random.choice(self._memory[symbol])
+                if symbol is None:
+                    break
+                chain.append(symbol)
+            yield chain
+
+
```
Comment (markzz) [inline jim/markov.py:28]: PEP-8: Too many blank lines.

File: jim/markov.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    def _get_starting_symbol(self):
+        if self._starting_symbols:
+            starting_symbol = random.choice(self._starting_symbols)
+            return random.choice(self._memory[starting_symbol])
+        else:
+            return random.choice(list(self._memory.keys()))            
+
+
```
Comment (markzz) [inline jim/markov.py:35]: PEP-8: Too many blank lines.

File: jim/markov.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    def memory_bytes(self):
+        return pickle.dumps(self._memory)
+
+
```
Comment (markzz) [inline jim/markov.py]: PEP-8: Too many blank lines.

File: jim/markov.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    def _fetch_pair(self, symbols):
+        for i in range(0, len(symbols) - 1):
+            yield symbols[i], symbols[i + 1]
+
+
```
Comment (markzz) [inline jim/markov.py]: PEP-8: Too many blank lines.

**Review by markzz — CHANGES_REQUESTED** [review]

File: jim/cmd_funcs.py (hunk 1 above)
Comment (Gagogaa) [reply jim/cmd_funcs.py]: I only see these variables being useful in this file so I don't see a need to move them to `util.py`. However they shouldn't be public. So I'll rename them to `_STORY_TIME_MARKOV` and `_READ_BOOKS`.

Comment (markzz) [reply jim/cmd_funcs.py]: Hmm, maybe perhaps have it separated out into the markov module similar to the DB module?

File: jim/markov.py [hunk 12]
```diff
@@ -24,26 +23,24 @@ def chains(self):
...
 
     def _get_starting_symbol(self):
         if self._starting_symbols:
             starting_symbol = random.choice(self._starting_symbols)
+
             return random.choice(self._memory[starting_symbol])
         else:
-            return random.choice(list(self._memory.keys()))            
-
+            return random.choice(list(self._memory.keys()))
```
Comment (markzz) [inline jim/markov.py:34]: Style: this particular return was fine. I only put in comments where I had issues with it.

File: jim/markov.py [hunk 13]
```diff
@@ -24,26 +23,24 @@ def chains(self):
...
                     break
                 chain.append(symbol)
-            yield chain
 
+            yield chain
 
     def _get_starting_symbol(self):
         if self._starting_symbols:
             starting_symbol = random.choice(self._starting_symbols)
+
```
Comment (markzz) [inline jim/markov.py]: Style: this particular return was fine. I only put in comments where I had issues with it.

**Review by markzz — CHANGES_REQUESTED** [review]

File: jim/markov.py (hunk 12 above)
Comment (Gagogaa) [reply jim/markov.py:34]: I'm using a new text editor and accidentally left white space at the end of the line in the previous commit

**Comment by markzz** [comment]
merge with changes needed from maintainer.

**Pull request merged**